### PR TITLE
Feature/timeseries category

### PR DESCRIPTION
**Pull request by adamkorynta**

### CODE REVIEW

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationCatalogEndpointInput.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    @Override
+    protected HttpRequestBuilder addInputParameters(HttpRequestBuilder httpRequestBuilder) {
+        String pageSizeString = Optional.ofNullable(pageSize).map(Object::toString).orElse(null);
+        return httpRequestBuilder.addQueryParameter("office", officeId)
+                                 .addQueryParameter("unitSystem", unitSystem)
+                                 .addQueryParameter("cursor", cursor)
+                                 .addQueryParameter("pageSize", pageSizeString)
+                                 .addQueryParameter("like", locationIdFilter)
```
Comment (rma-psmorris) [inline cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationCatalogEndpointInput.java:83]: how are the like parameters matched against? using oracle notation? glob/wildcard?

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationCatalogEndpointInput.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+ * SOFTWARE.
+ */
+
+package mil.army.usace.hec.cwms.radar.client.controllers;
+
+import java.util.Optional;
+import mil.army.usace.hec.cwms.http.client.EndpointInput;
+import mil.army.usace.hec.cwms.http.client.HttpRequestBuilder;
+
+public class LocationCatalogEndpointInput extends EndpointInput {
```
Comment (rma-psmorris) [inline cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationCatalogEndpointInput.java:31]: Is this a refactor change? I thought location catalog was already in another PR

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/CatalogController.java [hunk 3]
```diff
@@ -28,17 +28,26 @@
...
+import mil.army.usace.hec.cwms.radar.client.model.LocationCatalog;
 import mil.army.usace.hec.cwms.radar.client.model.RadarObjectMapper;
 import mil.army.usace.hec.cwms.radar.client.model.TimeSeriesCatalog;
 
 public final class CatalogController {
 
     private static final String CATALOG_TIMESERIES_ENDPOINT = "catalog/timeseries";
+    private static final String CATALOG_LOCATIONS_ENDPOINT = "catalog/locations";
 
     public TimeSeriesCatalog retrieveTimeSeriesCatalog(ApiConnectionInfo apiConnectionInfo, TimeSeriesCatalogEndpointInput input) throws IOException {
```
Comment (rma-psmorris) [inline cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/CatalogController.java:40]: why do you have location and time series catalogs here, but break out the other catalogs into separate classes?

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryController.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        String endpoint = timeSeriesEndpointInput.getCategoryId()
+                                                 .map(c -> TIME_SERIES_CATEGORY_ENDPOINT + "/" + c)
+                                                 .orElse(TIME_SERIES_CATEGORY_ENDPOINT + "/null");
+        HttpRequestResponse response = new HttpRequestBuilder(apiConnectionInfo, endpoint)
+            .addEndpointInput(timeSeriesEndpointInput)
+            .execute();
+        return RadarObjectMapper.mapJsonToObject(response.getBody(), TimeSeriesCategory.class);
+    }
+
+    List<TimeSeriesCategory> retrieveTimeSeriesCategories(ApiConnectionInfo apiConnectionInfo, TimeSeriesCategoryEndpointInput input)
```
Comment (rma-psmorris) [inline cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryController.java:50]: Why is this not public

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryEndpointInput.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        this.officeId = officeId;
+        return this;
+    }
+
+    Optional<String> getCategoryId() {
+        return Optional.ofNullable(categoryId);
+    }
+
+    @Override
+    protected HttpRequestBuilder addInputParameters(HttpRequestBuilder httpRequestBuilder) {
```
Comment (rma-psmorris) [inline cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryEndpointInput.java:57]: Where is the category id?

File: cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationsCatalogController.java [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+import java.io.IOException;
+import java.util.List;
+import mil.army.usace.hec.cwms.http.client.ApiConnectionInfo;
+import mil.army.usace.hec.cwms.http.client.ServerNotFoundException;
+import mil.army.usace.hec.cwms.radar.client.model.LocationAlias;
+import mil.army.usace.hec.cwms.radar.client.model.LocationCatalog;
+import mil.army.usace.hec.cwms.radar.client.model.LocationCatalogEntry;
+import org.junit.jupiter.api.Test;
+
+class TestLocationsCatalogController extends TestController {
```
Comment (rma-psmorris) [inline cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationsCatalogController.java:42]: Isnt this class in a different PR?

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryEndpointInput.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+package mil.army.usace.hec.cwms.radar.client.controllers;
+
+import java.util.Optional;
+import mil.army.usace.hec.cwms.http.client.EndpointInput;
+import mil.army.usace.hec.cwms.http.client.HttpRequestBuilder;
+
+/**
+ *
+ */
+public final class TimeSeriesCategoryEndpointInput extends EndpointInput {
```
Comment (rma-psmorris) [inline cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryEndpointInput.java:34]: unit test?

**Review by rma-psmorris — CHANGES_REQUESTED** [review]
First review

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationCatalogEndpointInput.java (hunk 2 above)
Comment (adamkorynta) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationCatalogEndpointInput.java]: I needed to rebase onto main. Fixed now

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/CatalogController.java (hunk 3 above)
Comment (adamkorynta) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/CatalogController.java]: Pretty much just to follow the query schema. I'm open to moving them into the TimeSeriesController and LocationsController classes.

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryController.java (hunk 4 above)
Comment (adamkorynta) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryController.java:50]: mistake

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryEndpointInput.java (hunk 5 above)
Comment (adamkorynta) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryEndpointInput.java:57]: It's a URL parameter, not a query parameter (not sure if I have the right terms)

http://localhost:11524/cwms-data/timeseries/category/{category-id}?office=SWT

File: cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationsCatalogController.java (hunk 6 above)
Comment (adamkorynta) [reply cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestLocationsCatalogController.java]: Fixed

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryEndpointInput.java (hunk 7 above)
Comment (adamkorynta) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryEndpointInput.java:34]: added

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationCatalogEndpointInput.java (hunk 1 above)
Comment (adamkorynta) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/LocationCatalogEndpointInput.java]: From the documentation: "Posix regular expression matching against the id"

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryEndpointInput.java (hunk 5 above)
Comment (rma-psmorris) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryEndpointInput.java:57]: That seems odd... What if the category doesnt exist?

File: cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestTimeSeriesCategoryEndpointInput.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    @Test
+    void testTimeSeriesCategoryEndpointInputGetAll() {
+        TimeSeriesCategoryEndpointInput timeSeriesCategoryEndpointInput = new TimeSeriesCategoryEndpointInput();
+        assertFalse(timeSeriesCategoryEndpointInput.getCategoryId().isPresent());
+    }
+
+    @Test
+    void testTimeSeriesCategoryEndpointInputCategoryId() {
+        TimeSeriesCategoryEndpointInput timeSeriesCategoryEndpointInput = new TimeSeriesCategoryEndpointInput("category-id");
+        assertEquals("category-id", timeSeriesCategoryEndpointInput.getCategoryId().get());
```
Comment (rma-psmorris) [inline cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestTimeSeriesCategoryEndpointInput.java:43]: no test on office id

**Review by rma-psmorris — CHANGES_REQUESTED** [review]
further review

File: cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestTimeSeriesCategoryEndpointInput.java (hunk 8 above)
Comment (adamkorynta) [reply cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestTimeSeriesCategoryEndpointInput.java:43]: Need to wait for Bryson's mock work to test that

File: cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryEndpointInput.java (hunk 5 above)
Comment (adamkorynta) [reply cwms-radar-client/src/main/java/mil/army/usace/hec/cwms/radar/client/controllers/TimeSeriesCategoryEndpointInput.java:57]: Then you get a 404 return code. This is a pretty standard REST design

File: cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestTimeSeriesCategoryEndpointInput.java (hunk 8 above)
Comment (adamkorynta) [reply cwms-radar-client/src/test/java/mil/army/usace/hec/cwms/radar/client/controllers/TestTimeSeriesCategoryEndpointInput.java:43]: https://jira.hecdev.net/browse/CWMS-687

**Comment by rma-psmorris** [comment]
no further review

**Review by rma-psmorris — APPROVED** [review]

**Pull request merged**